app/models/bookmark.rb

```diff
@@ -12,6 +12,17 @@ class Bookmark < ApplicationRecord
   validates_length_of :bookmarker_notes,
     maximum: ArchiveConfig.NOTES_MAX, too_long: ts("must be less than %{max} letters long.", max: ArchiveConfig.NOTES_MAX)
 
+  validate :not_already_bookmarked_by_user, on: :create
+  def not_already_bookmarked_by_user
+    return unless self.pseud && self.bookmarkable
+
+    bookmarks = Bookmark.join_bookmarkable.where(bookmarkable_id: self.bookmarkable_id)
```

Comment:
I think there might be an issue here. Bookmarkables can be of different types of objects (Work, Series, ExternalWork), which means we could have as many as three bookmarkables with the same id. That could give us false positives, e.g. if the user has bookmarked a Series with the id 123 and is now trying to bookmark a Work with the id 123, they might be told they have already bookmarked the item.

I'm also not sure if it's necessary to look at all of the bookmarks for a given bookmarkable. Would it instead make sense to do something like `pseud.user.bookmarks.where(bookmarkable: bookmarkable)` to look through the existing bookmarks for the user who owns the pseud? Or is the join just more efficient?

Reply:
I have no idea how to check efficiency of that code, but I highly doubt that join is more efficient in that case.
From the search size perspective it should make no difference if we go users-who-bookmarked-the-work or bookmarks-which-belong-to-user.
Changed.

-----------
This comment seems like a good occasion to ask for a bit of help :)

I struggled to find a way how to reach bookmarks-made-by-user. In the end I found similar code in [Gift](https://github.com/otwcode/otwarchive/blob/master/app/models/gift.rb#L31)  and copied it. Apparently it was not the best decision 😄 
I also was sure that as Bookmarkable is the parent "import" class (can't find what's is called properly in Ruby) of bookmarkable objects, the ids must be unique.

So I clearly misunderstand a lot :)

My problem is that I do not know how to get the full picture of classes, attributes, methods and relationships in the model:
- Visualizing the schema with rails-erd wasn't much help as result is either too simple or too convoluted
- I can't find any auto-docs
- Searching schema.rb clearly didn't help me see that there is a bookmarks attribute found on User :D 
- Simply reading through model classes doesn't help me much because of implicit attributes and relationships
- Ruby extension for vscode doesn't seem to be of help for that problem   

So I wanted to ask if there is any obvious way to create a full picture of the model or if it just must come from experience? :)

Reply:
>In the end I found similar code in Gift and copied it. Apparently it was not the best decision

That's a great way to approach a problem if you don't know what else to do, imo! It won't always give you the best solution, but it will at least give you a starting point, and then someone can help you refine it during code review.

>So I wanted to ask if there is any obvious way to create a full picture of the model or if it just must come from experience? :)

I will poke some folks and see if they have practical advice for you, since I do a lot of relying on familiarity with the code and site functionality. (For questions like "How do I get all of a user's bookmarks?", it might be a good idea to read over the [Active Record Associations docs](https://guides.rubyonrails.org/association_basics.html) if you haven't already, particularly the `has_one` and `has_many` sections.)